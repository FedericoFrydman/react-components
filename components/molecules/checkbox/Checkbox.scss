@import '../../assets/styles/colors.scss';

.lsux-cb-container {
  align-items: center;
  display: flex;
  margin: 2px;

  & input {
    height: 0;
    margin: 0;
    opacity: 0;
    width: 0;
  }

  body.using-key input:focus ~ &__cb {
    outline: 2px solid var(--checkbox-cb-enabled-focus-halo);
  }

  &__cb {
    align-items: center;
    border: 2px solid var(--checkbox-cb-enabled-unchecked-border);
    border-radius: 3px;
    cursor: pointer;
    display: flex;
    height: 22px;
    justify-content: center;
    min-width: 22px;
    transition: all 0.1s ease-in-out;

    img {
      display: none;
      height: 18px;
      user-select: none;
      width: 18px;
    }

    &--filled--error,
    &--outlined--error {
      border: 2px solid $R200;
    }
  }

  input:checked ~ &__cb {
    background-color: var(--checkbox-cb-enabled-checked-filled-background);
    border: 2px solid var(--brand-color-prim200);

    &--filled {
      background-color: var(--checkbox-cb-enabled-checked-filled-background);
      border-color: var(--checkbox-cb-enabled-checked-filled-border);
      img {
        display: block;
        filter: $filter-white;
      }
      &--error {
        background-color: $R200;
        border: 2px solid $R200;

        img {
          display: block;
          filter: $filter-white;
        }
      }
    }

    &--outlined {
      background-color: var(--checkbox-cb-enabled-checked-outlined-background);
      img {
        display: block;
        filter: var(--brand-color-prim200-filter);
      }
      &--error {
        background-color: $N00;
        border: 2px solid $R200;
        img {
          display: block;
          filter: $filter-red-200;
        }
      }
    }
  }

  input:disabled ~ &__cb {
    border: 2px solid $N500;
    cursor: not-allowed;
  }
  input:disabled:checked ~ &__cb {
    cursor: not-allowed;

    &--filled,
    &--filled--error {
      background-color: var(--checkbox-cb-disabled-checked-filled-background);
      border-color: var(--checkbox-cb-disabled-checked-filled-border);
      img {
        display: block;
        filter: $filter-white;
      }
    }

    &--outlined,
    &--outlined--error {
      background-color: var(--checkbox-cb-disabled-checked-outlined-background);
      border-color: var(--checkbox-cb-disabled-checked-outlined-border);
      img {
        display: block;
        filter: $filter-gray;
      }
    }

    img {
      display: block;
      filter: $filter-gray;
    }
  }

  input:disabled ~ .lsux-text {
    color: $N500;
    cursor: not-allowed;
  }
}
