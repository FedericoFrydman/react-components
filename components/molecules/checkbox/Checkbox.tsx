import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import { Icon } from '../../../components/atoms/icon/Icon';
import { Text } from '../../../components/atoms/text/Text';
import './Checkbox.scss';

export type CheckboxVariant = 'filled' | 'outlined';

export interface CheckboxProps extends Omit<React.HTMLProps<HTMLInputElement>, 'css' | 'name'> {
  /**
   * Determines if input is checked or not.
   * @default false
   */
  checked?: boolean;
  /**
   * Determines if the input can be of mixed state ('minus' icon when checked).
   * @default false
   */
  mixed?: boolean;
  /**
   * Determines if the checkbox displays an error state.
   * @default false
   */
  error?: boolean;
  /**
   * Determines if the checkbox is disabled or not.
   * @default false
   */
  disabled?: boolean;
  /**
   * Left label text for the checkbox.
   */
  leftLabel?: string;
  /**
   * Right label text for the checkbox.
   */
  rightLabel?: string;
  /**
   * Style of the checkbox. Either filled or outlined.
   * @default filled
   */
  variant?: CheckboxVariant;
  /**
   * Name of the input.
   */
  name: string;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * Additional data that is dispatched with the tracking event.
   */
  onClick?: (e: React.MouseEvent) => void | unknown;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLInputElement>;
}

const CheckboxComponent = (
  {
    checked = false,
    classNames = [],
    disabled = false,
    error = false,
    leftLabel,
    mixed = false,
    name,
    rightLabel,
    onClick,
    variant = 'filled',
    ...props
  }: CheckboxProps,
  ref: CheckboxProps['ref'],
) => {
  const classCheckboxContainer = 'lsux-cb-container';
  const classCheckboxElement = `${classCheckboxContainer}__cb`;
  const cnCheckbox = cn(classCheckboxElement, {
    [`${classCheckboxElement}--mixed`]: mixed,
    [`${classCheckboxElement}--${variant}`]: variant,
    [`${classCheckboxElement}--${variant}${error ? '--error' : ''}`]: variant,
    ['ml-3']: leftLabel,
    ['mr-3']: rightLabel,
  });

  return (
    <label onClick={onClick} className={cn('lsux-cb-container', classNames)} data-testid="lsux-cb-container">
      <input
        {...props}
        onClick={(e) => e.stopPropagation()}
        ref={ref}
        type="checkbox"
        checked={checked}
        name={name}
        disabled={disabled}
        className="lsux-cb-container__input"
      />
      {leftLabel && <Text text={leftLabel} />}
      <div className={cnCheckbox} data-testid="lsux-cb-container__icon">
        {checked && <Icon name={mixed ? 'edit_remove_minus' : 'interface_check'} alt={name} size="large" />}
      </div>
      {rightLabel && <Text text={rightLabel} />}
    </label>
  );
};

export const Checkbox = forwardRef<HTMLInputElement, CheckboxProps>(CheckboxComponent);
