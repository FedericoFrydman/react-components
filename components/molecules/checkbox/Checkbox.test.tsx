import { createRef } from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { render, screen } from '@testing-library/react';
import cn from 'classnames';

import { Checkbox, CheckboxProps, CheckboxVariant } from './Checkbox';

describe('Checkbox Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Checkbox name="checkbox-test" />);
    testId = 'test checkbox';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the checkbox the provided classNames', () => {
          const checkbox = renderCheckboxContainer({ classNames });

          expect(checkbox).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });

    describe('checked', () => {
      describe('when checked is provided', () => {
        it('gives the checkbox the checked property', () => {
          const checkbox = renderCheckbox({ checked: true });

          expect(checkbox).toHaveProperty('checked');
        });
      });
    });

    describe('leftLabel', () => {
      const label = 'my button label';

      it('is displayed when present', () => {
        renderCheckbox({ leftLabel: label });

        expect(screen.getByText(label)).toBeVisible();
      });
    });

    describe('rightLabel', () => {
      const label = 'my button label';

      it('is displayed when present', () => {
        renderCheckbox({ rightLabel: label });

        expect(screen.getByText(label)).toBeVisible();
      });
    });

    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the checkbox the provided classNames', () => {
          const checkbox = renderCheckboxContainer({ classNames });

          expect(checkbox).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });

    describe('left label', () => {
      const leftLabel = 'my checkbox leftLabel';

      it('is displayed when present', () => {
        renderCheckbox({ leftLabel });

        expect(screen.getByText(leftLabel)).toBeVisible();
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLInputElement>();

    it('provides a ref to the checkbox', () => {
      expect(renderCheckbox({ ref })).toEqual(ref.current);
    });
  });

  describe('unspecified html attributes', () => {
    const altText = 'My accessible checkbox';

    it('allows html attributes to be passed to the underlying checkbox element', () => {
      render(<Checkbox name={testId} alt={altText} data-testid={testId} />);

      const checkbox = screen.getByTestId(testId);

      expect(checkbox).toHaveAttribute('alt', altText);
    });
  });

  describe('variant', () => {
    const variants: CheckboxVariant[] = ['filled', 'outlined'];

    it(`defaults to 'filled'`, () => {
      const checkbox = renderCheckboxIcon();

      expect(checkbox).toHaveClass(`lsux-cb-container__cb--filled`, { exact: false });
    });

    function itAssignsTheRightClassForVariant(variant: CheckboxVariant) {
      const checkboxTypeClass = `lsux-cb-container__cb--${variant}`;

      describe(`when the variant is ${variant}`, () => {
        it(`assigns the ${checkboxTypeClass} to the Checkbox`, () => {
          const checkbox = renderCheckboxIcon({ checked: true, variant });

          expect(checkbox).toHaveClass(checkboxTypeClass);
        });
      });
    }

    variants.forEach((variant) => itAssignsTheRightClassForVariant(variant));
  });

  const renderCheckboxContainer = (props: Omit<CheckboxProps, 'name'> = {}): HTMLElement => {
    render(<Checkbox {...props} name={testId} />);

    return screen.queryByTestId('lsux-cb-container');
  };

  const renderCheckbox = (props: Omit<CheckboxProps, 'name'> = {}): HTMLElement => {
    render(<Checkbox {...props} name={testId} data-testid={testId} />);

    return screen.getByTestId(testId);
  };

  const renderCheckboxIcon = (props: Omit<CheckboxProps, 'name'> = {}): HTMLElement => {
    render(<Checkbox {...props} name={testId} data-testid={testId} />);

    return screen.getByTestId('lsux-cb-container__icon');
  };
});
