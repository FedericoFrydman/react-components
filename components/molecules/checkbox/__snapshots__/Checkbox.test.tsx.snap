// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox Component should render the component 1`] = `
<label
  className="lsux-cb-container"
  data-testid="lsux-cb-container"
>
  <input
    checked={false}
    className="lsux-cb-container__input"
    disabled={false}
    name="checkbox-test"
    onClick={[Function]}
    type="checkbox"
  />
  <div
    className="lsux-cb-container__cb lsux-cb-container__cb--filled"
    data-testid="lsux-cb-container__icon"
  />
</label>
`;
